### ISSUE TITLE
ICE on stable 1.23 during constant evaluation

### ISSUE THREAD

**Issue by matklad** [rust-lang/rust#47852]
Reproduction (not-minimized)

```
$ git clone https://github.com/matklad/exonum.git --branch ICE 
$ cd exonum/exonum
$ rustc --version
rustc 1.23.0 (766bd11c8 2018-01-01)
$ cargo check 

error: internal compiler error: /checkout/src/librustc_const_eval/pattern.rs:283: encountered errors lowering pattern: [ConstEval(ConstEvalErr { span: src/messages/protocol.rs:41:37: 41:58, kind: TypeckError })]
   --> src/messages/mod.rs:208:17
    |
208 |                 CONNECT_MESSAGE_ID => Any::Connect(Connect::from_raw(raw)?),
    |                 ^^^^^^^^^^^^^^^^^^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.23.0 (766bd11c8 2018-01-01) running on x86_64-unknown-linux-gnu

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:418:8
```

**Comment by matklad** [comment]
I think the following line is to blame: https://github.com/matklad/exonum/blob/d7d6a20ca176efb3c1379f9d64a2d25b660291fe/exonum/src/messages/protocol.rs#L41

It should be an error, because the `message_id` function is not const fn

**Comment by oli-obk** [comment]
I think this'll be fixed by #46882, too

**Comment by matklad** [comment]
Indeed it is! I get this correct error with nightly:

```
error[E0015]: calls in constants are limited to constant functions, struct and enum constructors
  --> src/messages/protocol.rs:41:37
   |
41 | pub const CONNECT_MESSAGE_ID: u16 = Connect::message_id();
   |                                     ^^^^^^^^^^^^^^^^^^^^^

```

👍 ❤️
